src/rt/cast_.d

```diff
@@ -19,6 +19,15 @@ extern (C):
 nothrow:
 pure:
 
+// because using == does a dynamic cast, but we
+// are trying to implement dynamic cast.
+extern (D) private bool areClassInfosEqual(scope const ClassInfo a, scope const ClassInfo b) @safe
+{
+    if (a is b)
+        return true;
+    return (a && b) && a.name == b.name;
```

Comment:
Not sure
* whether the null-checks are really required here, or are better moved to the few call sites potentially passing a null,
* whether the extra name check should be restricted to shared druntime (`*.{so,dll}`) only, in order not to slow down dynamic casts with static druntime, assuming static druntime isn't used for multiple D binaries in a process (with Windows exception for DMD) - see https://github.com/dlang/druntime/pull/2828#discussion_r689227660.